lib/adapters/class-adapter.php

```diff
@@ -36,20 +59,396 @@ abstract class Adapter {
+								case 'year':
+									$query_vars = [
+										'year' => gmdate( 'Y', $timestamp ),
+									];
+									$name       = gmdate( 'Y', $timestamp );
+									break;
+
+								case 'month':
+									$query_vars = [
+										'year'     => gmdate( 'Y', $timestamp ),
+										'monthnum' => gmdate( 'n', $timestamp ),
+									];
+									$name       = gmdate( 'F Y', $timestamp );
+									break;
+
+								case 'day':
+									$query_vars = [
+										'year'     => gmdate( 'Y', $timestamp ),
+										'monthnum' => gmdate( 'n', $timestamp ),
+										'day'      => gmdate( 'j', $timestamp ),
+									];
+									$name       = gmdate( 'F j, Y', $timestamp );
+									break;
+
+								default:
+									continue 3; // switch() is considered a looping structure.
+							}
+
+							break;
+
+						default:
+							// continue 2; // switch() is considered a looping structure.
+					}
+
+					$datum = [
+						'query_vars' => $query_vars,
+						'name'       => $name,
+						'count'      => $item['doc_count'],
+						'selected'   => $selected,
+					];
+				}
+
+				$facet_data[ $label ]->items[] = $datum;
+			}
+		}
+
+		return apply_filters( 'elasticsearch_extensions_facet_data', $facet_data );
+	}
+
+	/**
+	 * Get Facet by field
+	 *
+	 * @param string $field Facet field. See get_facet_data for acceptable values.
+	 * @param string $value Value corresponding to the field.
+	 * @return Facet|null
+	 */
+	public function get_facet_data_by( string $field = '', string $value = '' ) {
+		$facet_data = $this->get_facet_data();
+		foreach ( $facet_data as $facet ) {
+			if ( isset( $facet[ $field ] ) && $value === $facet[ $field ] ) {
+				return $facet;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Get the query var for a given taxonomy name.
+	 *
+	 * @access protected
+	 *
+	 * @param  string $taxonomy_name A valid taxonomy.
+	 * @return string The query var for the given taxonomy.
+	 */
+	protected function get_taxonomy_query_var( string $taxonomy_name ) {
+		$taxonomy = get_taxonomy( $taxonomy_name );
+
+		if ( ! $taxonomy || is_wp_error( $taxonomy ) ) {
+			return false;
+		}
+
+		return $taxonomy->query_var;
+	}
+
+	/**
+	 * Enables an aggregation based on WP Category.
+	 */
+	public function enable_category_aggregation(): void {
+		$this->aggregate_categories = true;
+	}
+
+	/**
+	 * Enables faceting on empty search query strings.
+	 */
+	public function enable_empty_search_faceting(): void {
+		$this->empty_search_faceting = true;
+	}
+
+	/**
+	 * Enables an aggregation based on post type.
+	 */
+	public function enable_post_date_aggregation(): void {
+		$this->aggregate_post_dates = true;
+	}
+
 	/**
 	 * Enables an aggregation based on post type.
 	 */
 	public function enable_post_type_aggregation(): void {
 		$this->aggregate_post_types = true;
 	}
 
+	/**
+	 * Enables an aggregation based on WP Tags.
+	 */
+	public function enable_tag_aggregation(): void {
```

Comment:
Is this because WordPress handles categories and tags differently in terms of what it calls them in query construction? Since we need to be able to enable aggregations on taxonomies generally (to support custom taxonomies) I'm wondering whether it makes sense to have categories and tags be treated in a "special" way here.